add initial 'data' to plugin context

arcadianCdr opened the pull request:
adds 'data' in plugin context based on pv's object data for first plugin run (was previously empty).

boi123212321 commented on src/plugins/events/actor.ts:
```diff
@@ -23,7 +24,21 @@ export async function onActorCreate(
   const createdImages = [] as Image[];
 
-  const pluginResult = await runPluginsSerial(config, event, {
+  const labels = (await Actor.getLabels(actor)).map((l) => l.name);
```
Just return the labels, not just their names

boi123212321 commented on src/plugins/events/actor.ts:
```diff
@@ -23,7 +24,21 @@ export async function onActorCreate(
+    bookmark: actor.bookmark ? actor.bookmark : undefined,
+    nationality: actor.nationality ? actor.nationality : undefined,
+    aliases: actor.aliases?.length ? actor.aliases : undefined,
+    labels: labels?.length ? labels : undefined,
```
To make it more consistent, just use labels, that way it's always an array

boi123212321 commented on src/plugins/events/actor.ts:
```diff
@@ -23,7 +24,21 @@ export async function onActorCreate(
+    favorite: actor.favorite,
+    bookmark: actor.bookmark ? actor.bookmark : undefined,
+    nationality: actor.nationality ? actor.nationality : undefined,
+    aliases: actor.aliases?.length ? actor.aliases : undefined,
```
Same as labels

boi123212321 commented on src/plugins/events/actor.ts:
```diff
@@ -23,7 +24,21 @@ export async function onActorCreate(
+  const labels = (await Actor.getLabels(actor)).map((l) => l.name);
+  const initialData: Dictionary<unknown> = {
+    name: actor.name,
+    description: actor.description ? actor.description : undefined,
```
I don't really understand the switch to undefined here, why not just use description?
Same for all the other properties

boi123212321 commented on src/plugins/events/scene.ts:
```diff
@@ -61,7 +62,28 @@ export async function onSceneCreate(
+    labels: labels?.length ? labels : undefined,
+    studio: scene.studio ? (await Studio.getById(scene.studio))?.name : undefined,
+    // If more than one movie, uses the first one
+    movie: (await Movie.getByScene(scene._id))?.[0]?.name,
```
Just return all movies, let the plugin use the first one if it wants to

boi123212321 commented on src/plugins/index.ts:
```diff
@@ -38,11 +39,12 @@ export async function runPluginsSerial(
         ...inject,
         pluginArgs,
       });
+      initialData = undefined;
```
Why?

boi123212321 commented on src/plugins/index.ts:
```diff
@@ -38,11 +39,12 @@ export async function runPluginsSerial(
     try {
       const pluginResult = await runPlugin(config, pluginName, {
-        data: <typeof result>JSON.parse(JSON.stringify(result)),
+        data: initialData ?? <typeof result>JSON.parse(JSON.stringify(result)),
```
This way only the first plugin has access to the data, maybe it should be provided under a different property, that way any plugin can differentiate between piped (new) data, and the initial data

arcadianCdr commented:
Modified according to our discussions to:
- Server functions to get the extra initial data: lazy loaded on first use (an result reused for subsequent calls as they are immutable).
- Removed compatibility with the plugin's output types to allow the full details of the object properties and the extra server functions => removed the initialData structure as they would have been a duplicate with the full initial object that is already passed to the context.
- As a help to plugin developers, I added a new getMergedData() method on the plugin side. It converts the initial data to the output type and merge with data (also in output type).
